File: jcs-vectors/python/check.py
#!/usr/bin/env python3
"""Python leg of the ATX v1.1 JCS byte-agreement gate.

For each vector under ../vectors, independently recompute the canonical bytes
from the vector's `tbs` object using the vendored RFC 8785 serializer, then
assert they reproduce the pinned `expected.canonicalHex` byte-for-byte. The
Python leg does not verify signatures (the post-quantum ML-DSA-65 library
landscape in Python is fragmented); the Go leg covers signature KATs. Agreement
on the canonical bytes is the property this leg proves.

Exit 0 iff every vector reproduces its pinned hex. Run via ../run-agreement.sh.
"""

from __future__ import annotations

import json
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent))
from rfc8785 import canonicalize  # noqa: E402


def vectors_dir() -> Path:
    here = Path(__file__).resolve().parent
    for cand in (here.parent / "vectors", here / "vectors"):
        if cand.is_dir():
            return cand
    sys.stderr.write("[py] error: could not locate vectors/ dir\n")
    sys.exit(2)


def main() -> int:
    vdir = vectors_dir()
    files = sorted(p for p in vdir.glob("*.json"))
    if not files:
        sys.stderr.write(f"[py] error: no vector files in {vdir}\n")
        return 2

    hex_by_name: dict[str, str] = {}
    npass = nfail = 0
    for path in files:
        vec = json.loads(path.read_text())
        name = vec["name"]
        want = vec["expected"]["canonicalHex"]
        got = canonicalize(vec["tbs"]).hex()
        hex_by_name[name] = got
        if got == want:
            print(f"[py]   PASS  {name:<34}  {len(got)//2} bytes")
            npass += 1
        else:
            print(f"[py]   FAIL  {name:<34}  canonical hex mismatch")
            print(f"       want: {want}")
            print(f"       got:  {got}")
            nfail += 1

    a, b = hex_by_name.get("key-order-scramble"), hex_by_name.get("baseline")
    if a is not None and b is not None:
        if a == b:
            print(f"[py]   PASS  {'invariant:scramble==baseline':<34}  identical canonical bytes")
        else:
            print(f"[py]   FAIL  {'invariant:scramble==baseline':<34}  scramble differs from baseline")
            nfail += 1

    print(f"\n[py]     {npass} pass, {nfail} fail ({len(files)} vectors)")
    return 0 if nfail == 0 else 1


if __name__ == "__main__":
    sys.exit(main())

File: jcs-vectors/ts/check.mjs
// TypeScript/JS leg of the ATX v1.1 JCS byte-agreement gate.
//
// For each vector under ../vectors, independently recompute the canonical bytes
// from the vector's `tbs` object using erdtman/canonicalize (the RFC 8785
// reference implementation, the same library the Secretless broker ATX verifier
// will use), then assert they reproduce the pinned expected.canonicalHex
// byte-for-byte. Agreement on the canonical bytes is the property this leg
// proves; signature verification is covered by the Go leg.
//
// Exit 0 iff every vector reproduces its pinned hex. Run via ../run-agreement.sh.

import { readFileSync, readdirSync } from "node:fs";
import { fileURLToPath } from "node:url";
import { dirname, join } from "node:path";
import { Buffer } from "node:buffer";
import canonicalize from "canonicalize";

const here = dirname(fileURLToPath(import.meta.url));
const vdir = join(here, "..", "vectors");

const files = readdirSync(vdir)
  .filter((f) => f.endsWith(".json"))
  .sort();

if (files.length === 0) {
  process.stderr.write(`[ts] error: no vector files in ${vdir}\n`);
  process.exit(2);
}

const hexByName = {};
let npass = 0;
let nfail = 0;

for (const f of files) {
  const vec = JSON.parse(readFileSync(join(vdir, f), "utf8"));
  const name = vec.name;
  const want = vec.expected.canonicalHex;
  // canonicalize() returns the RFC 8785 canonical UTF-8 string; encode to bytes
  // then hex so the comparison is over exact bytes, not JS string identity.
  const canonicalStr = canonicalize(vec.tbs);
  const got = Buffer.from(canonicalStr, "utf8").toString("hex");
  hexByName[name] = got;
  if (got === want) {
    console.log(`[ts]   PASS  ${name.padEnd(34)}  ${got.length / 2} bytes`);
    npass++;
  } else {
    console.log(`[ts]   FAIL  ${name.padEnd(34)}  canonical hex mismatch`);
    console.log(`       want: ${want}`);
    console.log(`       got:  ${got}`);
    nfail++;
  }
}

const a = hexByName["key-order-scramble"];
const b = hexByName["baseline"];
if (a !== undefined && b !== undefined) {
  if (a === b) {
    console.log(`[ts]   PASS  ${"invariant:scramble==baseline".padEnd(34)}  identical canonical bytes`);
  } else {
    console.log(`[ts]   FAIL  ${"invariant:scramble==baseline".padEnd(34)}  scramble differs from baseline`);
    nfail++;
  }
}

console.log(`\n[ts]     ${npass} pass, ${nfail} fail (${files.length} vectors)`);
process.exit(nfail === 0 ? 0 : 1);
